### PR TITLE
Feature/add coloring graph simulation

### PR DESCRIPTION
**Pull request by ChanaRoss**
i have created the graph coloring problem.
in addition i wrote a simple random policy which chooses an available node randomly and then chooses an optional color (new color/ color that has been used and not used by one of the Neighbors)
still need to add more tests.

### CODE REVIEW

File: src/graph_coloring_simulation/gc_simulation/problem_generator.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    :param probability_of_edge: Probability of an edge given any two different nodes
+    :type probability_of_edge: float
+    :return: (number of edges, adjacency list)
+    :rtype: (int, list of lists)
+    """
+
+    g = nx.fast_gnp_random_graph(number_of_nodes, probability_of_edge, seed=None, directed=False)
+    edges = []
+    for i in range(number_of_nodes):
+        temp1 = g.adj[i]
```
Comment (joeloren) [inline src/graph_coloring_simulation/gc_simulation/problem_generator.py:39]: unused

File: src/graph_coloring_simulation/gc_simulation/simulator.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        """
+        Create a new graph_coloring_simulation. Note that you need to call reset() before starting the cvrp_simulation.
+        :param num_nodes: number of nodes in the graph [int]
+        :param problem_generator: a generator of type ScenarioGenerator which generates one instance of the cvrp problem
+        and returns the initial state of the problem
+
+        """
+        super().__init__()
+        # initial state is empty data variables if self.reset() is not called
+        self.initial_state: State = State(unique_colors=set(), graph=nx.Graph(), num_colored_nodes=0, nodes_order=[])
```
Comment (joeloren) [inline src/graph_coloring_simulation/gc_simulation/simulator.py:39]: Where is the number of allowed node colours defined? I can't seem to find it.

File: src/graph_coloring_simulation/gc_simulation/simulator.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+                raise ValueError(f"chose invalid color for node id: {node_chosen}, "
+                                 f"problem is with neighbor node id:{neighbor_node}")
+        # assuming we didn't fail in the previous check, update the color of the node chosen to be the new color
+        self.current_state.graph.nodes[node_chosen]['color'] = color_chosen
+        if color_chosen not in self.current_state.unique_colors:
+            self.current_state.unique_colors.add(color_chosen)
+        self.current_state.num_colored_nodes += 1
+        self.current_state.nodes_order.append(node_chosen)
+        reward = len(self.current_state.unique_colors)
+        is_done = self.calc_is_done()
```
Comment (joeloren) [inline src/graph_coloring_simulation/gc_simulation/simulator.py:102]: By defining the terminal condition this way, you're requiring the graph to be fully k-colourable, which may not always be the case (for 3-colouring in fact, determining that a graph is 3-colourable is NP complete). An alternative approach is the optimization problem of maximizing the number of coloured nodes. So the stopping condition would be that there aren't any nodes that can be validly coloured.

**Review by joeloren — COMMENTED** [review]
Left a few questions in the code.

File: src/graph_coloring_simulation/gc_simulation/simulator.py (hunk 2 above)
Comment (ChanaRoss) [reply src/graph_coloring_simulation/gc_simulation/simulator.py:39]: the problem here doenst have a maximum number of colors allowed. you need to use the minimum number of colors possible. in the worst case the maximum number of colors will be the number of nodes in the graph

File: src/graph_coloring_simulation/gc_simulation/problem_generator.py (hunk 1 above)
Comment (ChanaRoss) [reply src/graph_coloring_simulation/gc_simulation/problem_generator.py:39]: true, i left this for future use.

**Pull request merged**